src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeWriter.java

```diff
@@ -0,0 +1,766 @@
+ * balanced tree.
+ * <p>
+ * Recursively builds a block balanced tree to assign all incoming points to smaller
+ * and smaller rectangles (cells) until the number of points in a given
+ * rectangle is &lt;= <code>maxPointsInLeafNode</code>.  The tree is
+ * fully balanced, which means the leaf nodes will have between 50% and 100% of
+ * the requested <code>maxPointsInLeafNode</code>.  Values that fall exactly
+ * on a cell boundary may be in either cell.
+ *
+ * <p>
+ * <b>NOTE</b>: This can write at most Integer.MAX_VALUE * <code>maxPointsInLeafNode</code> total points.
+ */
+public class BlockBalancedTreeWriter
+{
+    // Enable to check that values are added to the tree in correct order and within bounds
+    private static final boolean DEBUG = false;
+
+    // Default maximum number of point in each leaf block
+    public static final int DEFAULT_MAX_POINTS_IN_LEAF_NODE = 1024;
+
+    private final int bytesPerValue;
+    private final BytesRef scratchBytesRef1 = new BytesRef();
+    private final int maxPointsInLeafNode;
+    private final byte[] minPackedValue;
+    private final byte[] maxPackedValue;
+    private long pointCount;
+    private final long maxDoc;
+
+    public BlockBalancedTreeWriter(long maxDoc, int bytesPerValue, int maxPointsInLeafNode)
+    {
+        if (maxPointsInLeafNode <= 0)
+            throw new IllegalArgumentException("maxPointsInLeafNode must be > 0; got " + maxPointsInLeafNode);
+        if (maxPointsInLeafNode > ArrayUtil.MAX_ARRAY_LENGTH)
+            throw new IllegalArgumentException("maxPointsInLeafNode must be <= ArrayUtil.MAX_ARRAY_LENGTH (= " +
+                                               ArrayUtil.MAX_ARRAY_LENGTH + "); got " + maxPointsInLeafNode);
+
+        this.maxPointsInLeafNode = maxPointsInLeafNode;
+        this.bytesPerValue = bytesPerValue;
+        this.maxDoc = maxDoc;
+
+        minPackedValue = new byte[bytesPerValue];
+        maxPackedValue = new byte[bytesPerValue];
+    }
+
+    public long getPointCount()
+    {
+        return pointCount;
+    }
+
+    public int getBytesPerValue()
+    {
+        return bytesPerValue;
+    }
+
+    public int getMaxPointsInLeafNode()
+    {
+        return maxPointsInLeafNode;
+    }
+
+    /**
+     * Write the point values from a {@link IntersectingPointValues}. The points can be reordered before writing
+     * to disk and does not use transient disk for reordering.
+     */
+    public long writeField(IndexOutput out, IntersectingPointValues reader,
+                           final Callback callback) throws IOException
+    {
+        SAICodecUtils.writeHeader(out);
+
+        // We are only ever dealing with one dimension, so we can sort the points in ascending order
+        // and write out the values
+        if (reader.needsSorting())
+            MutablePointsReaderUtils.sort(Math.toIntExact(maxDoc), bytesPerValue, reader, 0, Math.toIntExact(reader.size()));
```

Comment:
I love that `sort()` here immediately subtracts 1 from `maxDoc`, and that's the only reason we can't just pass the max row ID, but have to first add one to it :D

Reply:
I have been looking at the current Lucene implementation and a lot of this weirdness has been fixed on their trunk. Specifically, they have changed the format of the tree when it is built so they don't need to then do the `rotateToTree` to rebuild it in order to write it to disk. These changes are primarily for the one-dimensional case so are very relevant to us. I am thinking that some of these changes could be pulled in as part of a separate ticket for phase 2.

Reply:
There's the temptation to get format changes in before our first release, but I know we've got to draw the line somewhere. It's probably okay for us to have a short-term focus, given the timeline for 5.0, and punt that to phase 2, as long as that Jira clearly describes what we might want to do so it isn't super foreign when we do pick it up.